core/commonMain/src/implementations/immutableList/PersistentVectorBuilder.kt

```diff
@@ -366,6 +667,254 @@ class PersistentVectorBuilder<E>(private var vector: PersistentList<E>,
+
+        // skip unaffected leaves
+        while (bufferSize == MAX_BUFFER_SIZE && leafIterator.hasNext()) {
+            bufferSize = removeAll(predicate, leafIterator.next(), MAX_BUFFER_SIZE, bufferRef)
+        }
+
+        // if no elements from the root match the predicate, check the tail
+        if (bufferSize == MAX_BUFFER_SIZE) {
+            assert(!leafIterator.hasNext())
+            val newTailSize = removeAllFromTail(predicate, tailSize, bufferRef)
+            if (newTailSize == 0) {
+                // all elements of the tail was removed, pull the last leaf from the root to make it the tail
+                pullLastBufferFromRoot(root, size, rootShift)
+            }
+            return newTailSize != tailSize
+        }
+
+        // handle affected leaves reusing mutable ones
+        val unaffectedElementsCount = leafIterator.previousIndex() shl LOG_MAX_BUFFER_SIZE
+
+        val buffers = mutableListOf<Array<Any?>>()
+        val recyclableBuffers = mutableListOf<Array<Any?>>()
+
+        while (leafIterator.hasNext()) {
+            val leaf = leafIterator.next()
+            bufferSize = recyclableRemoveAll(predicate, leaf, MAX_BUFFER_SIZE, bufferSize, bufferRef, recyclableBuffers, buffers)
+        }
+
+        // handle the tail
+        val newTailSize = recyclableRemoveAll(predicate, tail, tailSize, bufferSize, bufferRef, recyclableBuffers, buffers)
+
+        @Suppress("UNCHECKED_CAST")
+        val newTail = bufferRef.value as Array<Any?>
+        newTail.fill(null, newTailSize, MAX_BUFFER_SIZE)
+
+        // build the root
+        val newRoot = if (buffers.isEmpty()) root!! else placeBuffers(root, unaffectedElementsCount, rootShift, buffers.iterator())
+        val newRootSize = unaffectedElementsCount + (buffers.size shl LOG_MAX_BUFFER_SIZE)
+
+        root = retainFirst(newRoot, newRootSize)
+        tail = newTail
+        size = newRootSize + newTailSize
+
+        return true
+    }
+
+    /**
+     * Retains first [size] elements of the [root].
+     *
+     * If the height of the root is bigger than needed to store [size] elements, it's decreased.
+     */
+    private fun retainFirst(root: Array<Any?>, size: Int): Array<Any?>? {
+        check(size and MAX_BUFFER_SIZE_MINUS_ONE == 0)
+
+        if (size == 0) {
+            rootShift = 0
+            return null
+        }
+
+        val lastIndex = size - 1
+        var newRoot = root
+        while (lastIndex shr rootShift == 0) {
+            rootShift -= LOG_MAX_BUFFER_SIZE
+            @Suppress("UNCHECKED_CAST")
+            newRoot = root[0] as Array<Any?>
+        }
+        return nullifyAfter(newRoot, lastIndex, rootShift)
+    }
+
+    /**
+     * Nullifies nodes cells after the specified [index].
+     *
+     * Used to prevent memory leaks after reusing nodes.
+     */
+    private fun nullifyAfter(root: Array<Any?>, index: Int, shift: Int): Array<Any?> {
+        check(shift >= 0)
+
+        if (shift == 0) {
+            return root
```

Comment:
Need a comment explaining why `root` is expected to have nulls in this case.

Reply:
Added a comment explaining why leaf buffers are not filled with nulls